oak/server/time/roughtime_client.cc

```diff
@@ -0,0 +1,241 @@
+    close(handle);
+    return Status(asylo::error::GoogleError::INTERNAL,
+                  "Network error on sending request to " + server.name);
+  }
+
+  uint8_t receive_buffer[kReceiveBufferSize];
+  ssize_t receive_size;
+  do {
+    receive_size = recv(handle, receive_buffer, kReceiveBufferSize, 0 /* flags */);
+  } while (receive_size == -1 && errno == EINTR);
+  close(handle);
+  if (receive_size == -1) {
+    if (errno == EINTR) {
+      return Status(asylo::error::GoogleError::INTERNAL, "Request timed out for " + server.name);
+    }
+    return Status(asylo::error::GoogleError::INTERNAL, "No response received from " + server.name);
+  }
+
+  return std::string(reinterpret_cast<const char*>(receive_buffer),
+                     static_cast<size_t>(receive_size));
+}
+
+StatusOr<RoughtimeInterval> GetIntervalFromServer(const RoughtimeServerSpec& server) {
+  oak::NonceGenerator<roughtime::kNonceLength> generator;
+  std::string response;
+  Nonce<roughtime::kNonceLength> nonce = generator.NextNonce();
+  ASYLO_ASSIGN_OR_RETURN(response, SendRequest(server, nonce));
+  roughtime::rough_time_t timestamp_microseconds;
+  uint32_t radius_microseconds;
+  std::string error;
+  std::string public_key;
+  if (!absl::Base64Unescape(server.public_key_base64, &public_key)) {
+    return Status(asylo::error::GoogleError::INVALID_ARGUMENT,
+                  absl::StrFormat("Public key for server %s is not a valid base64 encoding.",
+                                  server.name.c_str()));
+  }
+
+  if (!roughtime::ParseResponse(&timestamp_microseconds, &radius_microseconds, &error,
+                                reinterpret_cast<const uint8_t*>(public_key.data()),
+                                reinterpret_cast<const uint8_t*>(response.data()), response.size(),
+                                nonce.data())) {
+    return Status(asylo::error::GoogleError::INTERNAL,
+                  absl::StrFormat("Response from %s could not be parsed: %s", server.name.c_str(),
+                                  error.c_str()));
+  }
+
+  if (radius_microseconds > kMaxRadiusMicroseconds) {
+    return Status(asylo::error::GoogleError::INTERNAL,
+                  absl::StrFormat("Radius from %s is too large: %" PRIu32 "", server.name.c_str(),
+                                  radius_microseconds));
+  }
+
+  if (radius_microseconds > timestamp_microseconds) {
+    return Status(asylo::error::GoogleError::INTERNAL,
+                  absl::StrFormat("Timestamp from %s is too small: %" PRIu64 "",
+                                  server.name.c_str(), timestamp_microseconds));
+  }
+
+  LOG(INFO) << "Time from " << server.name << ": " << timestamp_microseconds << " +/- "
+            << radius_microseconds;
+  return RoughtimeInterval{(timestamp_microseconds - radius_microseconds),
+                           (timestamp_microseconds + radius_microseconds)};
+}
+
+StatusOr<RoughtimeInterval> FindOverlap(const std::vector<RoughtimeInterval>& intervals,
```

Comment:
Could we get unit tests for this at some point (not necessarily now)?

Reply:
I am unsure of the best approach without unnecessarily increasing the visibility. Should I make this a private static method and use FRIEND_TEST, or should I move this to a separate library (eg roughtime_util.h) with only package level visibility? Or is there a better option?